staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go

```diff
@@ -188,8 +188,15 @@ func (c *DiscoveryController) sync(version schema.GroupVersion) error {
 		c.versionHandler.unsetDiscovery(version)
 		return nil
 	}
-	c.versionHandler.setDiscovery(version, discovery.NewAPIVersionHandler(Codecs, version, discovery.APIResourceListerFunc(func() []metav1.APIResource {
-		return apiResourcesForDiscovery
+
+	// Lazily calculate hash
+	hash := ""
+	c.versionHandler.setDiscovery(version, discovery.NewAPIVersionHandler(Codecs, version, discovery.APIResourceListerFunc(func() ([]metav1.APIResource, string) {
+		if len(hash) == 0 {
+			//!TODO: how to calculate hash?
+			hash = "calculatehash"
```

Comment:
My understanding is that this is the only place where we will have to calculate the hash for apiresources for CRDs, so however we do that is fine as long as the hash considers the entire content of the endpoint.

What about, mostly pseudocode:
```suggestion
			json, err := json.Marshal(apiResourcesForDiscovery)
			if err != nil {
				klog.Warnf("unable to marshal APIResources for CRD: %v", err)
				hash = ""
			} else {
				sum := md5.Sum([]byte(json))
				hash = hex.EncodeToString(sum[:])
			}
```